e2e/machineset_migration_test.go

```diff
@@ -213,13 +214,676 @@ var _ = Describe("[sig-cluster-lifecycle][OCPFeatureGate:MachineAPIMigration] Ma
+					[]*mapiv1beta1.MachineSet{mapiMachineSet},
+				)
+			})
+		})
+
+		Context("when MAPI MachineSet with spec.authoritativeAPI: MAPI and replicas 0", Ordered, func() {
+			It("should be rejected when scaling CAPI mirror", func() {
+				By("Scaling up CAPI MachineSet to 1 should be rejected")
+				capiframework.ScaleMachineSet(mapiMSAuthMAPIName, 1, capiframework.CAPINamespace)
+				mapiMachineSet, err := capiframework.GetMachineSet(cl, mapiMSAuthMAPIName, capiframework.CAPINamespace)
+				Expect(err).ToNot(HaveOccurred(), "Failed to get mapiMachineSet %s", mapiMachineSet)
+				Eventually(komega.Object(mapiMachineSet)).Should(HaveField("Spec.Replicas", HaveValue(Equal(int32(0)))))
+			})
+
+			It("should be rejected when updating CAPI mirror spec", func() {
+				By("Updating CAPI mirror spec (such as DeletePolicy)")
+				Eventually(func() error {
+					capiMachineSet, err = capiframework.GetMachineSet(cl, mapiMSAuthMAPIName, capiframework.CAPINamespace)
+					if err != nil {
+						return err
+					}
+					return komega.Update(capiMachineSet, func() {
+						capiMachineSet.Spec.DeletePolicy = "Oldest"
+					})()
+				}, capiframework.WaitMedium, capiframework.RetryShort).Should(Succeed(), "Failed to update CAPI MachineSet DeletePolicy")
+
+				By("Verifying both MAPI and CAPI MachineSet spec value are restored to original value")
+				Eventually(komega.Object(mapiMachineSet)).Should(HaveField("Spec.DeletePolicy", SatisfyAny(BeEmpty(), Equal("Random"))), "DeletePolicy should be either empty or 'Random'")
+				Eventually(komega.Object(capiMachineSet)).Should(HaveField("Spec.DeletePolicy", HaveValue(Equal("Random"))))
+			})
+
+			It("should create a new InfraTemplate when update MAPI MachineSet providerSpec", func() {
+				By("Updating MAPI MachineSet providerSpec InstanceType to m5.large")
+				newInstanceType := "m5.large"
+				originalAWSMachineTemplateName := capiMachineSet.Spec.Template.Spec.InfrastructureRef.Name
+				originalAWSMachineTemplate, err := capiframework.GetAWSMachineTemplateByPrefix(cl, mapiMSAuthMAPIName, capiframework.CAPINamespace)
+				Expect(err).ToNot(HaveOccurred(), "Failed to get original awsMachineTemplate  %s", originalAWSMachineTemplate)
+
+				patch := client.MergeFrom(mapiMachineSet.DeepCopy())
+				var awsProviderSpec mapiv1beta1.AWSMachineProviderConfig
+				Expect(json.Unmarshal(mapiMachineSet.Spec.Template.Spec.ProviderSpec.Value.Raw, &awsProviderSpec)).To(Succeed())
+				awsProviderSpec.InstanceType = newInstanceType
+				updatedSpec, err := json.Marshal(awsProviderSpec)
+				Expect(err).NotTo(HaveOccurred())
+				mapiMachineSet.Spec.Template.Spec.ProviderSpec.Value.Raw = updatedSpec
+				Expect(cl.Patch(ctx, mapiMachineSet, patch)).To(Succeed())
+
+				By("Waiting for new InfraTemplate to be created")
+				Eventually(func() bool {
+					capiMachineSet, err := capiframework.GetMachineSet(cl, mapiMSAuthMAPIName, capiframework.CAPINamespace)
+					if err != nil {
```

Comment:
Here were swallowing context too: if this fails in an e2e CI run, we will only see that `New InfraTemplate should be created` log line below.

  It would be great if we could leave breadcrumbs for debugging in here, e.g logging the errors